Weechat plugin split

lheckemann opened the pull request:
###### Motivation for this change
#22746 

This change moves each language plugin into a separate output, significantly reducing weechat's closure size. It also adds a `withPlugins` function to weechat, which generates a wrapper for adding the plugins.

This relies on weechat/weechat#971 and so probably shouldn't be merged until the next stable release which will (hopefully!) include that change.

This is probably my biggest nixpkgs change yet, so I'm PRing it now in search of feedback — any input would be highly appreciated! Particularly, the way in which I add the `withPlugins` function to weechat. This should allow using it like `weechat.withPlugins [weechat.ruby weechat.guile]`, but is there a more elegant way of adding it than the set union I'm doing here? I'd also like to know what people think about the mechanism as a whole, i.e. how this runtime composition is implemented.

Checklist before this can be merged:

- [x] ~~weechat/weechat#971 merged and incorporated into a release~~ patch added to expression instead
- [x] Remove references from main weechat derivation to python and perl packages — these should be in the wrapper as well
- [x] Document it — but where? This isn't specific to NixOS (so probably not the NixOS manual), nor is it only applicable to people intending to contribute to nixpkgs (so probably not the nixpkgs manual). Do we have an "end-user's guide" to nixpkgs?

Ideas:
 - [x] Potentially handle the plugins in a more unified way; I'll be experimenting with this a bit.
 - [ ] Allow managing user config as well as plugins


###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

lheckemann commented:
I would also appreciate if anyone could comment on whether they think https://github.com/lheckemann/nixpkgs/commit/e8bf1f9b9db78a3a10527d077f32dcfb0ffdfd6e is an improvement on this. It has yet to address the issue of plugin-specific environment variables properly, so is also WIP on top of this.

garbas commented:
@lheckemann to not wait for upstream you could also point to that PR and add it to the patches list.

lheckemann commented on pkgs/applications/networking/irc/weechat/default.nix:
```diff
@@ -18,64 +19,92 @@ assert tclSupport -> tcl != null;
+        platforms = stdenv.lib.platforms.unix;
+      };
+    };
+in weechat // {
```
I'm aware of this; the idea behind the approach I've used here is to take advantage of weechat's modularity to significantly reduce weechat's closure size without the need for rebuilding weechat itself (as is necessary if any of the `xSupport` flags are passed in). Customising weechat's plugins has already been possible simply by disabling them in the build, but that means that it needs to be rebuilt on the user's end which somewhat dampens the resulting improvement in closure size as all the build deps need to be available (at least initially) and the build takes longer.

garbas replied:
using ``withPlugins`` in does not go against reducing weechat's closure size. If you looked at neovim's expression you will see that by default default weechat is returned. if plugins are set you wrap the returning weechat executable. The only things that changes is how you go about customizing this. The preferred way to do this is via ``.override`` function.

lheckemann replied:
Right, that makes sense... so something like d421f4aefcaddf0ec07582ffb62fa6f625dd8f96 ?

(usage: `weechat.override {withPlugins = ["perl" "python" "ruby"];}`)

garbas replied:
👍

lheckemann commented:
Of course, I forgot about that!

garbas commented on pkgs/applications/networking/irc/weechat/default.nix:
```diff
@@ -18,64 +19,92 @@ assert tclSupport -> tcl != null;
+      src = fetchFromGitHub {
+        owner = "lheckemann";
+        repo = "weechat";
+        rev = "546e260fd58e41db20e7bc4ec5585e93fb6994c1";
```
i this v1.7.1 tag?

lheckemann replied:
No, it's master as of yesterday + my patch. I've changed it back to use v1.7.1 plus my patch
https://github.com/NixOS/nixpkgs/pull/25274/commits/368edaf44f4224e8699b9c66035716ae892d0fe2

garbas replied:
sorry commenting on old code :)

garbas reviewed (approved):
@lheckemann 💯 on the change. I would just squash the commits before we merge this to nixpkgs.

lheckemann commented:
Yes, that was my plan. I also still need to fix the perl/python path stuff in the wrapper — currently just `weechat` will still depend on perl and python, without even bringing in the plugins.

lheckemann commented:
It also still needs docs, which I don't know where to put.

lheckemann commented:
This definitely isn't ready for merging yet — for instance, `fish.py` from the official scripts repo doesn't work with it

garbas commented:
@lheckemann I think you can open a new section about weechat at [Chapter 10. Package Notes](http://nixos.org/nixpkgs/manual/#chap-package-notes)

I'll do a bit of dreaming in this comment and you stop me when I go too far :P Also I'm interested to work on this myself, just for some things I'm too short on my skills.

(1) nixos module: maybe it would be worthwhile to also create a nixos module for weechat. eg 

```
programs.weechat.enable = true;
programs.weechat.plugins = [ ... ];
programs.weechat.<other-options>...
```

(2) like with plugins it would be nice to have a variable to configure weechat to write logs in location other then ``$WEECHAT_HOME/logs``, we could add this as an argument like withPlugins

(3) configuring weechat: it would be also nice to be able to disallow weechat to write configuration. and create a nice way to configure weechat (eg. like i already do with neovim for example), we could add this as an argument like withPlugins

lheckemann commented:
1. Not sure how useful this would be, given that weechat doesn't really require any system-wide configuration — most modules, as far as I can tell, exist because they require system-wide settings e.g. `programs.zsh` for registering it as the users' default shell, `programs.cdemu` for setting appropriate permissions, etc. A module for weechat, as far as I can tell, wouldn't provide anything that installing an expression wouldn't already.

2. Yes.

3. Yes. Generally, separating the state and the config would be nice.

4. I'm not aware of any plugins that don't ship with weechat. Most extra functionality seems to be implemented through scripts built on the existing language plugins.

garbas commented:
1. you're probably right. ``.override`` to configure weechat is enough. maybe having a service that would spawn weechat+tmux like i do [here](https://github.com/garbas/dotfiles/blob/master/nixos/floki.nix#L105)

2. & 3. I would love to poke with weechat's code but this goes beyond my skills.

4. Sorry for the confusion. I was talking about scripts. instead of using weechat to pull in *unknown* code we can use nix. That would of-course require us to disable ``/scripts`` command.

lheckemann commented:
Yeah, I'll have a look at making those changes to weechat in the next few days.

As for script installation, yes I guess that would be the logical next step from there. I'll see what I can do!

lheckemann commented:
The python plugin needs some python libraries on its path. I think the right solution would be to allow passing in a python environment as created by `python3.withPackages`, but I'm not sure how best to expose this. @garbas do you have any ideas for what a nice way to do this would look like? Some ideas so far…

```nix
weechat.override { configure = {
  plugins = ["python"];
  environment = {
      PYTHON_HOME = python.withPackages (ps: with ps; [pycrypto dbus-python]);
  };
};}
```
But I'm not really sure if this is a nice way to do it, particularly because it relies on "magical" values so much. It would be nicer if it could somehow be included in the plugins attribute, something like

```nix
weechat.override { configure = {
  plugins = available: [
    (available.python.withPackages (ps: with ps; [pycrypto python-dbus]))
  ];
};}
```
or even
```nix
weechat.override { configure = {availablePlugins, ...}: {
  plugins = [availablePlugins.python.withPackages (ps: with ps; [pycrypto python-dbus])];
};}
```

But all of these look rather complicated to use. Any thoughts?

FRidh commented on pkgs/applications/networking/irc/weechat/default.nix:
```diff
@@ -1,81 +1,129 @@
+            pluginFile = "${weechat.python}/lib/weechat/plugins/python.so";
+            withPackages = pkgsFun: (python // {
+              extraEnv = ''
+                export PYTHONHOME="${pythonPackages.python.withPackages pkgsFun}"
```
This is not going to work. `python.withPackages` uses itself `--set PYTHONHOME`.

lheckemann replied:
It worked for me.

lheckemann replied:
What python.withPackages does is build an environment containing symlinks to all the packages *and* a python wrapper which sets `PYTHONHOME` to the location of the resulting environment. The wrapper won't actually be used by weechat's python plugin, but the python plugin does load the python interpreter as a shared library and at that point the interpreter looks at PYTHONHOME to see where to import stuff from.

FRidh commented on pkgs/applications/networking/irc/weechat/default.nix:
```diff
@@ -1,81 +1,129 @@
+        ln -s $plugin $out/plugins
+      done
+    '';
+  in writeScriptBin "weechat" ''
```
Have you considered using `buildEnv` for composition? This does look neat btw.

lheckemann replied:
Not sure how that would work in this case, since the paths are the actual shared libraries and not the derivations.

lheckemann commented:
>does the basic package, without plugins, need Python, or is it only needed for a plugin? If it does not build against a Python env, then you can just add the Python env to PATH.

The python plugin for weechat requires python obviously. It does not execute python as a process though, it uses the library interface to the interpreter, so I think the approach I'm taking is the right one?

lheckemann commented:
That's it documented. If nobody has any further objections, I'll rebase and squash for the actual merge.

0xABAB commented on doc/package-notes.xml:
```diff
@@ -652,4 +652,34 @@ cp ${myEmacsConfig} $out/share/emacs/site-lisp/default.el
+<section xml:id="sec-weechat">
+<title>Weechat</title>
+<para>
+Weechat can currently be configured to include your choice of plugins.
```
"You can configure plugins as follows" is shorter and says the same.

lheckemann replied:
Explicitly mentioning that this is an expression to install is more helpful, as some other packages are configured via nixpkgs config settings instead.

0xABAB commented on pkgs/applications/networking/irc/weechat/default.nix:
```diff
@@ -1,81 +1,129 @@
+        ++ extraBuildInputs;
+
+      NIX_CFLAGS_COMPILE = "-I${python}/include/${python.libPrefix}"
+        # Fix '_res_9_init: undefined symbol' error
```
Fixes?

lheckemann replied:
This is the same as before. I don't see a reason to change it.

0xABAB commented on pkgs/applications/networking/irc/weechat/default.nix:
```diff
@@ -1,81 +1,129 @@
+      postInstall = with stdenv.lib; ''
+        for p in ${concatMapStringsSep " " (p: p.name) enabledPlugins}; do
+          from=$out/lib/weechat/plugins/$p.so
+          to=''${!p}/lib/weechat/plugins/$p.so
```
These single quotes look out of place, but perhaps this is intentional.

lheckemann replied:
Yes. `${!p}` would be interpreted as an interpolation by nix and break it. `''${!p}` turns into `${!p}` which is passed on to bash, and expands to e.g. `$perl` which in turn is expanded into the path for the perl output.

0xABAB commented on pkgs/applications/networking/irc/weechat/default.nix:
```diff
@@ -1,81 +1,129 @@
-    + (stdenv.lib.optionalString stdenv.isDarwin "-DBIND_8_COMPAT=1 -lresolv");
+      meta = {
+        homepage = http://www.weechat.org/;
+        description = "A fast, light and extensible chat client";
```
I think it's an IRC client, because that's builtin.

lheckemann replied:
3 reasons not to changes this which would be each individually be enough standing alone IMO:

 - It supports other protocols such as XMPP through plugins/scripts.
 - The IRC plugin can be removed/excluded.
 - **This description is copied from [the website](https://weechat.org/).**

Consider looking stuff up if you're not even sure it's wrong rather than trying to correct me?

0xABAB commented:
I have no feedback on the actual features, so I will leave that to others. I am happy to see documentation again.

lheckemann commented:
@FRidh @garbas ping?

FRidh commented:
@lheckemann I'd like to give the final word to whoever maintains this

lheckemann commented:
Oh yeah, maybe @the-kenny and @lovek323 would like to weigh in as well :)

disassembler commented:
@lheckemann merge conflicts can you rebase? Also @the-kenny and @lovek323 we're looking for feedback here. If we don't get a response, I say we go ahead and merge it.

lheckemann committed:
weechat: split plugins into separate outputs

Also add a wrapper generator that allows adding the plugins back
conveniently and corresponding documentation in the package notes
section of the nixpkgs manual.

jluttine commented:
Would this PR make it possible to use plugins like https://github.com/wee-slack/wee-slack which, for instance, depend on other Python packages?

lheckemann commented:
@jluttine yes, see the addition to the manual for details on how. The packages in question do need to be included in nixpkgs though.

Ma27 commented:
@lheckemann are there any remaining blockers? I've packaged some Weechat plugins for Nixpkgs and I'd try to make them compatible with this in the next time :-)

lheckemann commented:
Just approval from maintainers.

Ma27 commented:
ah nice :-D
will try to make my weechat derivations compatible with this :-)

Ma27 commented:
@lheckemann just to ensure that I got this right: your change adds the ability to builda weechat with plugins.
My PR (see #28897) adds an FHS wrapper to make it possible that the scripts inside are buildable with nix. It surely needs some refactoring (I hacked this in ~1h to have an example that can be used for further discussion about this), but in the end it should live happily together with this, right?

lheckemann commented:
Yes, in principle: they're concerned with different parts of weechat, mine is for plugins and yours for scripts. I think the "patch weechat to search an alternative path" approach would be nicer than building a FHS env for scripts too, but I don't see any reason the latter approach would be incompatible with this PR either.

Ma27 commented:
just found your weechat patch. If such an approach is applicable to weechat scripts as well, I'll close this and open another PR 

> On 17. Nov 2017, at 10:32 AM, Linus Heckemann <notifications@github.com> wrote:
> 
> Yes, in principle: they're concerned with different parts of weechat, mine is for plugins and yours for scripts. I think the "patch weechat to search an alternative path" approach would be nicer than building a FHS env for scripts too, but I don't see any reason the latter approach would be incompatible with this PR either.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jluttine commented:
I started using this feature and it seems really great. I figured out how to add Python dependencies that some scripts need, but I couldn't figure out how to do the same for Lua dependencies.. There's no `lua.withPackages` apparently as there is for Python. Any ideas? I'm trying to add `cjson` Lua package required by this script: https://github.com/torhve/weechat-matrix-protocol-script

lheckemann commented:
Yes, it doesn't currently have support for this, largely because lua library support in nixpkgs isn't that great in general as far as I remember. I seem to remember @teto doing some work on that, maybe you can advise?

teto commented:
the work is visible here https://github.com/NixOS/nixpkgs/pull/33903. It works well except for some torch packages. I have yet to understand what's the problem there.

jluttine commented:
Ok, thanks! So at the moment, weechat cannot be used with lua dependencies. (?)

teto commented:
Other packages hack their way around via adjusting the LUA_PATH in a wrapper but if you are not hurry, there should be a better way soon :)
